### PR TITLE
feat(subscriptions): gate prompt guide behind its own feature flag

### PR DESCRIPTION
**Pull request by pauldambra**
i want to be able to release insight subscription AI summary separately to allowing users to add a prompt guide when they set a summary up

so a new flag

## Summary

Gates the \`summary_prompt_guide\` field — the per-subscription override that lets users steer the AI summary prompt — behind a new feature flag \`subscription-ai-summary-prompt-guide\`.

Kept separate from \`hackathons_subscriptions\` (which still gates the summary toggle + delivery history). This lets AI summaries ship publicly (flipping \`hackathons_subscriptions\` to 100%) **without** bundling in the \"customers can pass arbitrary instructions to the model\" decision.

## What changes

**Frontend** — nested \`FlaggedFeature\` around just the prompt-guide textarea, inside the existing HACKATHONS_SUBSCRIPTIONS block. Toggle stays visible; textarea appears only when both flags are on.

**Backend** — \`SubscriptionSerializer.validate\` rejects writes to \`summary_prompt_guide\` with \`403 PermissionDenied\` when the flag is off for the caller's organization. Three behaviours locked in by tests:
- **Field absent from payload** passes through unchanged, so unrelated PATCHes (e.g. \`{\"title\": \"...\"}\`) against a subscription with an existing prompt guide don't trip the gate.
- **Clearing the field** (\`\"\"\`) is always allowed so users aren't stuck with a value they can't edit if the flag flips off after they set one.
- **Non-empty writes** require the flag.

Flag key on both sides: \`subscription-ai-summary-prompt-guide\` (kebab-case, matches newer convention). Server-side evaluation scoped by organization via posthoganalytics, matching the pattern already used in \`ee/api/ee_event_definition.py\`.

## Test plan

- [x] 5 new tests in \`ee/api/test/test_subscription.py\`:
  - \`test_cannot_set_prompt_guide_when_feature_flag_disabled\` — create rejected (403)
  - \`test_can_set_prompt_guide_when_feature_flag_enabled\` — create accepted (201)
  - \`test_cannot_patch_prompt_guide_when_feature_flag_disabled\` — patch rejected (403)
  - \`test_can_clear_prompt_guide_when_feature_flag_disabled\` — clearing always works (200)
  - \`test_can_patch_unrelated_fields_with_existing_prompt_guide_when_flag_disabled\` — field-absent PATCH doesn't trip gate (200)
- [x] Existing 4 AI-consent tests still pass
- [x] \`ruff check\` and \`ruff format\` clean

## LLM context

For Paul: the frontend toggle behaviour is unchanged — \`HACKATHONS_SUBSCRIPTIONS\` still gates it. Just the textarea is now gated separately. Backend validation only fires when \`summary_prompt_guide\` is explicitly present in the payload, so nothing existing breaks. To release AI summaries: flip \`hackathons_subscriptions\` to 100%. The prompt guide stays hidden/rejected until you decide on it and flip \`subscription-ai-summary-prompt-guide\` separately.

---
*Created with [PostHog Code](https://posthog.com/code?ref=pr)*

### CODE REVIEW

File: ee/api/subscription.py [hunk 1]
```diff
@@ -189,9 +189,21 @@ def validate(self, attrs):
...
+        # "field absent" and "field set to null"; the `in attrs` check separates them so
+        # unrelated PATCHes don't trip the gate.
+        if "summary_prompt_guide" in attrs:
+            prompt_guide = attrs["summary_prompt_guide"]
+            if prompt_guide and len(prompt_guide) > 500:
+                raise ValidationError({"summary_prompt_guide": ["AI summary context must be 500 characters or fewer."]})
+            # Only gate on non-empty writes — clearing the field should always be allowed
+            # so users can't get stuck with a value they can no longer edit if the flag
+            # is flipped off after they set one.
+            if prompt_guide and not self._prompt_guide_feature_enabled():
```
Comment (pauldambra) [inline ee/api/subscription.py]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[convergent: qa-team/data-integrity + xp]** 🟠 HIGH

The flag is write-only — it gates `validate()` on create/PATCH but `snapshot_activities._sanitize_prompt_guide(subscription.summary_prompt_guide or "")` keeps passing the stored value to the LLM on every delivery regardless of flag state.

xp flagged the same concern from the UX side: the backend comment says "clearing is always allowed so users aren't stuck", but the frontend nests the textarea inside the new flag's `FlaggedFeature` — so when the flag flips off, users **can't see** the field to clear it. The escape hatch only reaches them if the UI still renders.

Net effect: flipping `hackathons_subscriptions` to 100% while leaving this flag off doesn't fully enforce "customers don't steer the prompt" — any sub that already has a guide stored (from the hackathon period) keeps steering it in prod.

Options, best → worst:
1. **Also gate the read-side** in `snapshot_activities.py` — when the flag is off for the subscription's org, pass `prompt_guide=""` to the LLM even if a value is stored. Closes both the data-integrity hole and the UX asymmetry in one place.
2. **Document** that the flag gates future writes only; existing stored values continue to take effect. Leaves the prompt-injection exposure open but at least matches reality.
3. **Clear stored values** on flag-flip. Destructive and weird.

File: ee/api/test/test_subscription.py [hunk 2]
```diff
@@ -468,6 +468,85 @@ def test_can_patch_unrelated_fields_when_summary_enabled_and_ai_consent_revoked(
         assert response.status_code == status.HTTP_200_OK
         assert response.json()["title"] == "Updated title"
 
+    def test_cannot_set_prompt_guide_when_feature_flag_disabled(self):
```
Comment (pauldambra) [inline ee/api/test/test_subscription.py]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[convergent: xp + qa-team/test-coverage]** 🟢 LOW

Five near-identical tests each setting `is_ai_data_processing_approved`, patching `feature_enabled`, calling `_create_subscription`, asserting status + body. Prime candidate for `@parameterized.expand` — CLAUDE.md explicitly prefers this — with a table of `(flag_state, payload, expected_status, expected_body_contains)`. Would also make the four-case matrix legible from one place rather than across five test methods.

Bonus: one extra case worth adding — explicit `{"summary_prompt_guide": null}` as a clear. The code comment distinguishes null from absent, but only `""` is tested.

File: ee/api/test/test_subscription.py [hunk 3]
```diff
@@ -468,6 +468,85 @@ def test_can_patch_unrelated_fields_when_summary_enabled_and_ai_consent_revoked(
...
+        value they can no longer edit if the flag is flipped off after they set one.
+        """
+        self.organization.is_ai_data_processing_approved = True
+        self.organization.save()
+
+        with patch("ee.api.subscription.posthoganalytics.feature_enabled", return_value=True):
+            create_response = self._create_subscription(summary_enabled=True, summary_prompt_guide="original")
+        subscription_id = create_response.json()["id"]
+
+        with patch("ee.api.subscription.posthoganalytics.feature_enabled", return_value=False):
```
Comment (pauldambra) [inline ee/api/test/test_subscription.py]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[xp]** ⚪ NIT

Test docstrings (`"Clearing the field must always be allowed..."`) directly contradict the project convention in CLAUDE.md: "Python tests: do not add doc comments." The behaviour is in the test name — if the name isn't carrying enough, rename the test.

File: ee/api/subscription.py [hunk 4]
```diff
@@ -189,9 +189,21 @@ def validate(self, attrs):
             if not allowed:
                 raise ValidationError({"target_value": [f"Invalid webhook URL: {error}"]})
 
-        prompt_guide = attrs.get("summary_prompt_guide")
-        if prompt_guide and len(prompt_guide) > 500:
-            raise ValidationError({"summary_prompt_guide": ["AI summary context must be 500 characters or fewer."]})
+        # Only validate / reject prompt-guide writes when the field is explicitly present
```
Comment (pauldambra) [inline ee/api/subscription.py]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[xp]** ⚪ NIT

Three names for one concept across two layers: UI label "Context for the AI summary", ValidationError "AI summary context", PermissionDenied "custom AI summary context". Pick one — probably "AI summary context" — and say it the same way everywhere the user could see it.

File: ee/api/subscription.py [hunk 5]
```diff
@@ -202,6 +214,28 @@ def validate(self, attrs):
...
+        members. `only_evaluate_locally=False` so we respect server-side cohort
+        / property conditions — this isn't on a hot path.
+        """
+        request = self.context.get("request")
+        if not request or not getattr(request, "user", None) or not getattr(request.user, "distinct_id", None):
+            return False
+        organization = self.context["get_organization"]()
+        org_id = str(organization.id) if organization else ""
+        return bool(
+            posthoganalytics.feature_enabled(
```
Comment (pauldambra) [inline ee/api/subscription.py:224]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[convergent: paul + qa-team/reliability + xp]** 🟢 LOW

`only_evaluate_locally=False` means every write that includes a non-empty `summary_prompt_guide` blocks on an external flag evaluation. The `if prompt_guide and ...` short-circuit correctly bounds this to actual-write paths (clears + absent PATCHes don't pay the cost), so blast radius is limited today.

Once `hackathons_subscriptions` hits 100% and this flag stays the only gate, the write-path still carries a network hop. Worth a follow-up to flip to local eval once the flag stabilises, or at least make sure latency on this serializer is visible. Not blocking.

File: ee/api/subscription.py [hunk 6]
```diff
@@ -202,6 +214,28 @@ def validate(self, attrs):
...
         return attrs
 
+    def _prompt_guide_feature_enabled(self) -> bool:
+        """Evaluate the prompt-guide feature flag for the caller's organization.
+
+        Scoped by organization (not user) so the gate is stable across a team's
+        members. `only_evaluate_locally=False` so we respect server-side cohort
+        / property conditions — this isn't on a hot path.
+        """
+        request = self.context.get("request")
```
Comment (pauldambra) [inline ee/api/subscription.py:218]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[paul]** 🟢 LOW

Fail-closed without a log means a misconfigured auth layer looks identical to "flag off" at 3am. Worth a debug log (`"denied prompt-guide write: no distinct_id on request"`) so you can tell them apart. Not blocking.

File: ee/api/subscription.py [hunk 7]
```diff
@@ -202,6 +214,28 @@ def validate(self, attrs):
...
+        if not request or not getattr(request, "user", None) or not getattr(request.user, "distinct_id", None):
+            return False
+        organization = self.context["get_organization"]()
```
Comment (pauldambra) [inline ee/api/subscription.py:221]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[paul]** ⚪ NIT

`org_id = str(organization.id) if organization else ""` → passing `""` to posthog feature_enabled silently runs the flag eval with no group context. In practice `get_organization()` always returns one on this path, so this branch is defensive-dead-code. I'd rather either assert/explode or pass no groups at all rather than an empty string that looks valid.

**Review by pauldambra — COMMENTED** [review]
🤖 QA Swarm review — 1 HIGH (convergent), 1 MEDIUM (convergent), 6 LOW/NIT. See inline comments.

File: ee/api/subscription.py (hunk 5 above)
Comment (pauldambra) [inline ee/api/subscription.py:224]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[qa-team/security]** 🟢 LOW

`posthoganalytics.feature_enabled(...)` can return `None` on evaluation errors (e.g. network timeout when `only_evaluate_locally=False`). `bool(None)` is `False`, so the gate safely denies — good. Worth a one-line test asserting deny-on-error (mock `feature_enabled` to return `None` and assert 403) so a future refactor can't accidentally invert this into fail-open.

**Comment by pauldambra** [comment]
> [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human
>
> Multi-perspective review: qa-team, paul-reviewer, xp-reviewer

## Verdict: ⚠️ REQUEST CHANGES

One material finding, flagged by two reviewers independently from different angles: **the flag is write-only** — existing stored prompt guides continue to flow into the LLM regardless of flag state, and the frontend hides the textarea when the flag is off so users can't even clear a stored value. Flipping `hackathons_subscriptions` to 100% while leaving this flag off would NOT fully enforce "customers don't steer the prompt." Rest are polish.

### Key findings

- 🟠 **HIGH — convergent (qa-team/data-integrity + xp)**: Flag is write-only. Stored `summary_prompt_guide` values keep being passed to the LLM in `snapshot_activities.py` on every delivery after the flag flips off. Also breaks the "clearing is always allowed" escape hatch because the frontend removes the textarea when the flag is off. Best fix: gate the read-side in the snapshot activity too.
- 🟢 **LOW — convergent (paul + qa-team + xp)**: `only_evaluate_locally=False` adds a network hop to the serializer write path. Bounded today by the `if prompt_guide and ...` short-circuit; worth a follow-up once the flag stabilises.
- 🟢 **LOW — convergent (qa-team + xp)**: Five nearly-identical tests should be parameterized (CLAUDE.md preference).
- ⚪ **NIT — xp**: Test docstrings violate the project's "no doc comments on Python tests" rule.
- ⚪ **NIT — xp**: Three names for the same concept across UI / ValidationError / PermissionDenied.
- ⚪ **NIT — paul**: Fail-closed path (no distinct_id / no org) should log so 3am-you can distinguish it from "flag off."
- 🟢 **LOW — qa-team/security**: Add a deny-on-error test (mock `feature_enabled` to return `None`, assert 403) so a future refactor can't invert into fail-open.

### Convergence

- **Read-side gate** — qa-team/data-integrity (HIGH) + xp (MEDIUM, UX angle). Same concern, two angles. Highest-confidence finding and the only one I'd block on.
- **`only_evaluate_locally=False`** — flagged by all three, all LOW. Follow-up material.
- **Parameterize tests** — qa-team/test-coverage + xp. Easy win, matches CLAUDE.md.

### Reviewer summaries

| Reviewer | Assessment |
| --- | --- |
| 🔍 qa-team | Mechanics are sound (safe-deny on missing distinct_id / flag-eval error, correct `in attrs` distinction). Material concern: flag is write-only, existing stored guides keep steering the LLM. |
| 👤 paul | LGTM shape — the split is exactly the right call. Small observability nits around the denial path, but nothing blocking. The backend 403 is distinguishable from the AI-consent 403, which matters at 3am. |
| 📐 xp | Simplicity rules mostly hold, but code leans on comments/docstrings to say what names and tests could say. Decide the UX product answer and make both sides say the same thing. |

---
*Automated by QA Swarm — not a human review*

**Comment by pauldambra** [comment]
* **#56225** <a href="https://app.graphite.com/github/pr/PostHog/posthog/56225?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/PostHog/posthog/56225?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: ee/api/subscription.py (hunk 1 above)
Comment (pauldambra) [reply ee/api/subscription.py]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `454fa9dbc60` — added a read-side gate in `snapshot_activities.py` via `_prompt_guide_feature_enabled_for_subscription`, plus two activity tests that assert the stored `summary_prompt_guide` is dropped when the flag is off and passed through when on. Flipping the flag off now stops stored values from steering the LLM on the very next run.

File: ee/api/test/test_subscription.py (hunk 2 above)
Comment (pauldambra) [reply ee/api/test/test_subscription.py]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `454fa9dbc60` — collapsed to a single `@parameterized.expand` table covering the four PATCH cases plus a new `deny_on_feature_flag_eval_error` row for the `None`-return safe-deny path.

File: ee/api/test/test_subscription.py (hunk 3 above)
Comment (pauldambra) [reply ee/api/test/test_subscription.py]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `454fa9dbc60` — test docstrings removed (CLAUDE.md says no docstrings on Python tests). The parameterized case names carry the intent.

File: ee/api/subscription.py (hunk 4 above)
Comment (pauldambra) [reply ee/api/subscription.py]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `454fa9dbc60` — backend 403 and 400 error bodies now both use "AI summary context" to match the UI label.

File: ee/api/subscription.py (hunk 5 above)
Comment (pauldambra) [reply ee/api/subscription.py:224]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

The `bool(...)` wrapper around `feature_enabled(...)` already converts a `None` return into `False`, so eval errors fail closed today. The new parameterized case `deny_on_feature_flag_eval_error` in `test_subscription.py` (added in `454fa9dbc60`) pins that behaviour by mocking `feature_enabled` to return `None` and asserting 403.

Comment (pauldambra) [reply ee/api/subscription.py:224]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Deferred — `only_evaluate_locally=False` is intentional while the flag is new and narrowly rolled out so we don't fail-closed on users in local-eval-only environments before the flag's rollout config has propagated. Once the flag stabilises we'll flip to `True` to cut the per-write round trip. Tracking as a follow-up, not blocking this PR.

File: ee/api/subscription.py (hunk 6 above)
Comment (pauldambra) [reply ee/api/subscription.py:218]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Deferred — agreed a debug log on the no-`distinct_id` fail-closed branch would help 3am triage. Skipping in this PR to keep the diff small; will add alongside the `only_evaluate_locally` flip when we tune the gate.

File: ee/api/subscription.py (hunk 7 above)
Comment (pauldambra) [reply ee/api/subscription.py:221]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Deferred — passing an empty `org_id` when `organization` is `None` is intentional: no real organization has that id, so PostHog's group evaluation can't match a cohort on it and the call fails-closed by default. Noting as a nit; happy to swap to an explicit early return if we touch this helper again.

File: frontend/src/lib/components/Subscriptions/views/EditSubscription.tsx [hunk 8]
```diff
@@ -449,12 +449,18 @@ export function EditSubscription({
...
-                                    />
-                                </LemonField>
+                                <FlaggedFeature flag={FEATURE_FLAGS.SUBSCRIPTION_AI_SUMMARY_PROMPT_GUIDE}>
+                                    <LemonField
+                                        name="summary_prompt_guide"
+                                        label="Context for the AI summary"
+                                        showOptional
+                                    >
+                                        <LemonTextArea
+                                            placeholder="e.g. This is a daily revenue health check - focus on revenue drop-off and churn signals"
```
Comment (MattPua) [inline frontend/src/lib/components/Subscriptions/views/EditSubscription.tsx:459]: Maybe instead of a placeholder  it would be better to have like a supplementary description field that gives different ideas of what would be good context.

Bullet pointed otherwise might confine too much the creativity of what this could be, if that makes sense

**Review by MattPua — APPROVED** [review]
